cleverhans/experimental/certification/dual_formulation.py

```diff
@@ -303,31 +357,88 @@ def get_full_psd_matrix(self):
         axis=0)
     return self.matrix_h, self.matrix_m
 
-  def compute_certificate(self):
+  def make_M_psd(self):
+    """Run binary search to find a value for nu that makes M PSD
+
+    Args:
+      feed_dict: dictionary of updated lambda variables to feed into M
+    Returns:
+      new_nu: new value of nu
+    """
+    original_nu = self.sess.run(self.nu)
+    _, min_eig_val_m = self.get_lanczos_eig()
+
+    lower_nu = original_nu
+    upper_nu = original_nu
+    num_iter = 0
+
+    # Find an upper bound on nu
+    while min_eig_val_m - TOL < 0:
+      if num_iter >= 15:
+        break
+      num_iter += 1
+      upper_nu *= 1.1
+      self.sess.run(tf.assign(self.nu, upper_nu))
+      _, min_eig_val_m = self.get_lanczos_eig()
+
+    final_nu = upper_nu
+
+    # Perform binary search to find best value of nu
+    while lower_nu <= upper_nu:
+      if num_iter >= 25:
+        final_nu = upper_nu
+        self.sess.run(tf.assign(self.nu, final_nu))
+        break
+      num_iter += 1
+      mid_nu = (lower_nu + upper_nu) / 2
+      self.sess.run(tf.assign(self.nu, mid_nu))
+      _, min_eig_val_m = self.get_lanczos_eig()
+      if min_eig_val_m - TOL < 0:
+        lower_nu = mid_nu
+      else:
+        upper_nu = mid_nu
+
+    # Add 0.05 to final nu to account for numerical instability
+    return original_nu, final_nu + 0.05
```

Comment:
Does it make sense to reassign value `original_nu` to the `self.nu`?
If not please update description of the function to mention that it has side effect by modifying nu

Reply:
i have a comment at the bottom of compute certificate for this